federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,842 @@
+package federatedtypes
+
+import (
+	"time"
+
+	"fmt"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	restclient "k8s.io/client-go/rest"
+	federationapi "k8s.io/kubernetes/federation/apis/federation/v1beta1"
+	federationclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset"
+	fedutil "k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+)
+
+const (
+	HpaKind              = "horizontalpodautoscaler"
+	HpaControllerName    = "horizontalpodautoscalers"
+	scaleForbiddenWindow = 5 * time.Minute
+)
+
+func init() {
+	RegisterFederatedType(HpaKind, HpaControllerName, []schema.GroupVersionResource{apiv1.SchemeGroupVersion.WithResource(HpaControllerName)}, NewHpaAdapter)
+}
+
+type HpaAdapter struct {
+	*schedulingAdapter
+	client federationclientset.Interface
+}
+
+func updateStatus(fedHpa *autoscalingv1.HorizontalPodAutoscaler, newStatus hpaFederatedStatus) (bool, autoscalingv1.HorizontalPodAutoscalerStatus) {
+	// Average out the available current utilisation
+	if *newStatus.currentCPUUtilizationPercentage != 0 && newStatus.count != 0 {
+		*newStatus.currentCPUUtilizationPercentage /= newStatus.count
+	}
+	gen := fedHpa.Generation
+	newFedHpaStatus := autoscalingv1.HorizontalPodAutoscalerStatus{ObservedGeneration: &gen}
+	needUpdate := false
+	if (fedHpa.Status.CurrentCPUUtilizationPercentage == nil &&
+		*newStatus.currentCPUUtilizationPercentage != 0) ||
+		(fedHpa.Status.CurrentCPUUtilizationPercentage != nil &&
+			*newStatus.currentCPUUtilizationPercentage !=
+				*fedHpa.Status.CurrentCPUUtilizationPercentage) {
+		needUpdate = true
+		newFedHpaStatus.CurrentCPUUtilizationPercentage = newStatus.currentCPUUtilizationPercentage
+	}
+	if (fedHpa.Status.LastScaleTime == nil && newStatus.lastScaleTime != nil) ||
+		(fedHpa.Status.LastScaleTime != nil && newStatus.lastScaleTime == nil) ||
+		((fedHpa.Status.LastScaleTime != nil && newStatus.lastScaleTime != nil) &&
+			newStatus.lastScaleTime.After(fedHpa.Status.LastScaleTime.Time)) {
+		needUpdate = true
+		newFedHpaStatus.LastScaleTime = newStatus.lastScaleTime
+	}
+	if fedHpa.Status.DesiredReplicas != newStatus.desiredReplicas {
+		needUpdate = true
+		newFedHpaStatus.CurrentReplicas = newStatus.currentReplicas
+	}
+	if fedHpa.Status.CurrentReplicas != newStatus.currentReplicas {
+		needUpdate = true
+		newFedHpaStatus.DesiredReplicas = newStatus.desiredReplicas
+	}
+	return needUpdate, newFedHpaStatus
+}
+
+func NewHpaAdapter(client federationclientset.Interface, config *restclient.Config) FederatedTypeAdapter {
+	schedulingAdapter := schedulingAdapter{
+		updateStatusFunc: func(obj pkgruntime.Object, schedulingInfo interface{}) error {
```

Comment:
@irfanurrehman I wonder if this doesn't make sense to have as an overridden method in the schedulingAdapter as well, if you're doing that for the other two methods. It seems odd that the HPA controller overrides two methods and implements this one here. It seems like it should either override all of the methods or none of them.

How much functionality in the schedulingAdapter is the HPA controller actually using? It seems to me like you should probably implement the interface directly, rather than including the schedulingAdapter as an embedded struct. Perhaps renaming schedulingAdapter to replicaSchedulingAdapter or something similar, so that it's clear that it's not a canonical implementation of the SchedulingAdapter interface, but the one that works for workload types that are only concerned with replicas.